docs/fn-2.0/README.md

```diff
@@ -0,0 +1,416 @@
+# Fn 2.0
+
+## NOTE: This is a proposal for updating Fn with the following goals:
+
+* Split HTTP routes from functions
+* Support CloudEvent messages as the native format throughout
+* Simplify everything
+* Modularize components
+* Namespacing
+
+## Removals
+
+* Default, HTTP and JSON formats will be removed
+* Cold functions will be removed
+* Async, queues are event sources now
+
+## Terminology
+
+Fn follows the terminology defined here: https://github.com/cloudevents/spec/blob/master/spec.md#notations-and-terminology
+
+As well as the following:
+
+### Event Sources
+
+Event sources are things that emit events and pass the event messages to Fn. Three common event sources are:
+
+* HTTP
+* Scheduler
+* Queue
+
+### Functions
+
+Functions are little programs that accept a defined input and output a defined output. They do not need to be aware of any other parts
+of the system and can run standalone; from the command line, etc. These can be written in any language and are packaged up as a container image. 
+
+* Functions read from STDIN, write to STDOUT and log to STDERR
+* Functions are packaged in container images
+* Input is a stream of CloudEvent messages in JSON format
+* Output is a stream of CloudEvent messages in JSON format
```

Comment:
We do a lot of construction in order to write back a JSON representation of the CloudEvent. Shouldn't protocol dispatcher be responsible for constructing the CloudEvent object from the function's response?